### ISSUE TITLE
Use users' credential providers for working with authenticated NuGet sources

### ISSUE THREAD

**Issue by mjrousos** [dotnet/upgrade-assistant#359]
Currently, Upgrade-Assistant uses NuGet.config files relevant to the project being upgraded to pull NuGet packages from the right feeds. But if any of the packages come from authenticated feeds, it can fail (with a FatalProtocolException; 401) due to not using available credential providers.

**Comment by MisinformedDNA** [comment]
Is there any workaround for this?

**Comment by grayjeremy** [comment]
Have you tried this?  https://github.com/microsoft/artifacts-credprovider
It's been a year or two since I've looked at it. (I'm not on the .net team, just a person with an idea)

**Comment by MisinformedDNA** [comment]
@grayjeremy I have it installed. It doesn't work.

**Comment by ahutch95** [comment]
Im also having this issue. Any solutions?

**Comment by sunandabalu** [comment]
StackTrace: 
```
System.Net.Http.HttpRequestException: Response status code does not indicate success: 401 (Unauthorized).
   at System.Net.Http.HttpResponseMessage.EnsureSuccessStatusCode()
   at NuGet.Protocol.HttpSource.<>c__DisplayClass15_0`1.<<GetAsync>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at NuGet.Common.ConcurrencyUtilities.ExecuteWithFileLockedAsync[T](String filePath, Func`2 action, CancellationToken token)
   at NuGet.Common.ConcurrencyUtilities.ExecuteWithFileLockedAsync[T](String filePath, Func`2 action, CancellationToken token)
   at NuGet.Protocol.HttpSource.GetAsync[T](HttpSourceCachedRequest request, Func`2 processAsync, ILogger log, CancellationToken token)
   at NuGet.Protocol.PackageMetadataResourceV3.LoadRegistrationIndexAsync(HttpSource httpSource, Uri registrationUri, String packageId, SourceCacheContext cacheContext, Func`2 processAsync, ILogger log, CancellationToken token)
   at NuGet.Protocol.PackageMetadataResourceV3.GetMetadataAsync(String packageId, Boolean includePrerelease, Boolean includeUnlisted, VersionRange range, SourceCacheContext sourceCacheContext, ILogger log, CancellationToken token)
   at NuGet.Protocol.PackageMetadataResourceV3.GetMetadataAsync(String packageId, Boolean includePrerelease, Boolean includeUnlisted, SourceCacheContext sourceCacheContext, ILogger log, CancellationToken token)
   at Microsoft.DotNet.UpgradeAssistant.MSBuild.PackageLoader.CallWithRetryAsync[T](Func`1 func)
   at Microsoft.DotNet.UpgradeAssistant.MSBuild.PackageLoader.GetLatestVersionAsync(String packageName, Boolean includePreRelease, String[] packageSources, CancellationToken token) in /_/src/components/Microsoft.DotNet.UpgradeAssistant.MSBuild/PackageLoader.cs:line 113
   at Microsoft.DotNet.UpgradeAssistant.Steps.Packages.Analyzers.UpgradeAssistantReferenceAnalyzer.AnalyzeAsync(IProject project, PackageAnalysisState state, CancellationToken token) in /_/src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/Analyzers/UpgradeAssistantReferenceAnalyzer.cs:line 50
   at Microsoft.DotNet.UpgradeAssistant.Steps.Packages.PackageUpdaterStep.RunPackageAnalyzersAsync(IUpgradeContext context, CancellationToken token) in /_/src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/PackageUpdaterStep.cs:line 206
   at Microsoft.DotNet.UpgradeAssistant.Steps.Packages.PackageUpdaterStep.InitializeImplAsync(IUpgradeContext context, CancellationToken token) in /_/src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/PackageUpdaterStep.cs:line 89
```

**Comment by mjrousos** [comment]
Looks like a bug in the package loader. I'll take a look as soon as I'm done with #367, probably end of this week.

**Comment by mjrousos** [comment]
Quick status update - I've managed to reproduce the issue locally and am investigating how we can use authentication plugins to avoid the error.

**Comment by thomas11** [comment]
@mjrousos I'm having trouble seeing how often upgrade-assistant is released or what the meaning of the version number is. How do I know when your commit makes it into a release?
Thanks for the quick fix!

**Comment by sunandabalu** [comment]
@thomas11 The best place to map a commit to a release is the changelog - https://github.com/dotnet/upgrade-assistant/blob/main/CHANGELOG.md. We will update the release version to changelog once we release. The typical cadence we have currently is once in 2 weeks.

**Comment by IcyCaster** [comment]
@thomas11 I too am eagerly waiting for this fix! 🤞 they release it soon!
(Thanks for the amazingly quick fix @mjrousos)

**Comment by sunandabalu** [comment]
This fix is released to nuget.org - https://www.nuget.org/packages/upgrade-assistant/0.2.222702

**Comment by thomas11** [comment]
I tried it but unfortunately I'm seeing no difference. Interactive mode is the default, correct? Do I need to do anything else to make it use the credential providers?

```
-----------------------------------------------------------------------------------------
- Microsoft .NET Upgrade Assistant v0.2.222702+aa9fd4e57ceef6ec3c09e585da50b2c0ba8986b4 -
-----------------------------------------------------------------------------------------

[16:50:19 INF] MSBuild registered from C:\Program Files\dotnet\sdk\5.0.202\
[16:50:19 INF] Registered 1 extensions:
        Default extension
[16:50:41 INF] Loading upgrade progress file at E:\AzureSearch\.upgrade-assistant
[16:50:41 INF] Initializing upgrade step Back up project
[16:50:41 INF] Initializing upgrade step Convert project file to SDK style
[16:50:41 INF] Initializing upgrade step Clean up NuGet package references
[16:51:02 INF] Marking package Microsoft.Spatial for removal because it appears to be a transitive dependency
[16:51:02 INF] Marking assembly reference System.configuration for removal based on package mapping configuration System.Configuration
[16:51:02 INF] Adding package System.Configuration.ConfigurationManager based on package mapping configuration System.Configuration
[16:51:02 FTL] Unexpected exception analyzing package references for: E:\AzureSearch\Source\Common\Product\Util\Util.csproj
NuGet.Protocol.Core.Types.FatalProtocolException: Unable to load the service index for source https://msdata.pkgs.visualstudio.com/bbd6e814-7e33-452b-9cc7-690473457fb8/_packaging/AzureSearch-Aggregated/nuget/v3/index.json.
 ---> System.Net.Http.HttpRequestException: Response status code does not indicate success: 401 (Unauthorized)
```

**Comment by mjrousos** [comment]
Yes, interactive mode should be the default. Can you try running with `-v` to get more details on the failure?

**Comment by thomas11** [comment]
This is the verbose output. Seems like it doesn't find a credential provider.

Before that there's a bunch of `[Warning] Problem loading file in MSBuild workspace Found project reference without a matching metadata reference`, not sure if that could have an impact.

```
[09:15:43 DBG] Analyzing packages with Duplicate reference analyzer
[09:15:44 DBG] Analyzing packages with Transitive reference analyzer
[09:15:44 DBG] Analyzing packages with Package map reference analyzer
[09:15:44 DBG] Looking for package maps in Default extension
[09:15:44 DBG] Loaded 4 package maps from C:\Users\thkapple\.dotnet\tools\.store\upgrade-assistant\0.2.222702\upgrade-assistant\0.2.222702\tools\net5.0\any\PackageMaps\AspNetPackageMap.json
[09:15:44 DBG] Loaded 3 package maps from C:\Users\thkapple\.dotnet\tools\.store\upgrade-assistant\0.2.222702\upgrade-assistant\0.2.222702\tools\net5.0\any\PackageMaps\AssemblyPackageMaps.json
[09:15:44 DBG] Loaded 7 package maps from C:\Users\thkapple\.dotnet\tools\.store\upgrade-assistant\0.2.222702\upgrade-assistant\0.2.222702\tools\net5.0\any\PackageMaps\DefaultPackageMap.json
[09:15:44 DBG] Finished loading package maps from extension Default extension
[09:15:44 DBG] Analyzing packages with Target compatibility reference analyzer
[09:15:44 DBG] NuGet package DynamicIssuerCertificateAuthorization, Version=3.13.1499 not found in package cache
[09:15:45 DBG] Using NuGet credential providers:
[09:15:45 FTL] Unexpected exception analyzing package references for: E:\AzureSearch\Source\Common\Product\Util\Util.csproj
NuGet.Protocol.Core.Types.FatalProtocolException: Unable to load the service index for source https://msdata.pkgs.visualstudio.com/bbd6e814-7e33-452b-9cc7-690473457fb8/_packaging/AzureSearch-Aggregated/nuget/v3/index.json.
 ---> System.Net.Http.HttpRequestException: Response status code does not indicate success: 401 (Unauthorized).
```

**Comment by mjrousos** [comment]
The warnings about the MSBuild workspace are probably unrelated. I agree that it looks like it's not finding the credential provider. This is what my output looks like:

```
[14:11:15 DBG] Analyzing packages with Upgrade assistant reference analyzer
[14:11:17 VRB] [NuGet] Using C:\Users\mikerou\.nuget\plugins\netcore\CredentialProvider.Microsoft\CredentialProvider.Microsoft.dll as a credential provider plugin.
[14:11:17 DBG] Using NuGet credential providers: SecurePluginCredentialProvider_C:\Users\mikerou\.nuget\plugins\netcore\CredentialProvider.Microsoft\CredentialProvider.Microsoft.dll
```

A couple follow-up questions -

1. What credential provider are you using? Maybe the [logic](https://github.com/dotnet/upgrade-assistant/blob/main/src/components/Microsoft.DotNet.UpgradeAssistant.MSBuild/NuGetCredentialsStartup.cs#L46) I added to register them isn't finding all relevant providers.
    1. I'll see if I can find more guidance on `CredentialService` and `ICredentialProvider` to make sure we're using them correctly.
2. Do you get any more of that callstack so that I can see where the HttpRequestException is coming from?

**Comment by mjrousos** [comment]
Looking at [how NuGet.exe gets credential providers](https://github.com/NuGet/NuGet.Client/blob/17c4f841ff61d27fe6b57cf45ceef16037062635/src/NuGet.Clients/NuGet.CommandLine/Commands/Command.cs#L209), there are a few differences. They have their own handlers for searching for plugin providers and prompting for credentials beyond what's exposed through `SecurePluginCredentialProviderBuilder`. I can try to replicate those behaviors in Upgrade Assistant.

**Comment by thomas11** [comment]
Here's the full stack trace. I'll look into whether we're doing anything special regarding Nuget credentials.

```[09:15:45 DBG] Using NuGet credential providers:
[09:15:45 FTL] Unexpected exception analyzing package references for: E:\AzureSearch\Source\Common\Product\Util\Util.csproj
NuGet.Protocol.Core.Types.FatalProtocolException: Unable to load the service index for source https://msdata.pkgs.visualstudio.com/bbd6e814-7e33-452b-9cc7-690473457fb8/_packaging/AzureSearch-Aggregated/nuget/v3/index.json.
 ---> System.Net.Http.HttpRequestException: Response status code does not indicate success: 401 (Unauthorized).
   at System.Net.Http.HttpResponseMessage.EnsureSuccessStatusCode()
   at NuGet.Protocol.HttpSource.<>c__DisplayClass15_0`1.<<GetAsync>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at NuGet.Common.ConcurrencyUtilities.ExecuteWithFileLockedAsync[T](String filePath, Func`2 action, CancellationToken token)
   at NuGet.Common.ConcurrencyUtilities.ExecuteWithFileLockedAsync[T](String filePath, Func`2 action, CancellationToken token)
   at NuGet.Protocol.HttpSource.GetAsync[T](HttpSourceCachedRequest request, Func`2 processAsync, ILogger log, CancellationToken token)
   at NuGet.Protocol.ServiceIndexResourceV3Provider.GetServiceIndexResourceV3(SourceRepository source, DateTime utcNow, ILogger log, CancellationToken token)
   --- End of inner exception stack trace ---
   at NuGet.Protocol.ServiceIndexResourceV3Provider.GetServiceIndexResourceV3(SourceRepository source, DateTime utcNow, ILogger log, CancellationToken token)
   at NuGet.Protocol.ServiceIndexResourceV3Provider.TryCreate(SourceRepository source, CancellationToken token)
   at NuGet.Protocol.Core.Types.SourceRepository.GetResourceAsync[T](CancellationToken token)
   at NuGet.Protocol.Core.Types.SourceRepository.GetResourceAsync[T]()
   at NuGet.Protocol.HttpFileSystemBasedFindPackageByIdResourceProvider.TryCreate(SourceRepository sourceRepository, CancellationToken token)
   at NuGet.Protocol.Core.Types.SourceRepository.GetResourceAsync[T](CancellationToken token)
   at Microsoft.DotNet.UpgradeAssistant.MSBuild.PackageLoader.GetPackageArchiveAsync(NuGetReference packageReference, CancellationToken token) in /_/src/components/Microsoft.DotNet.UpgradeAssistant.MSBuild/PackageLoader.cs:line 262
   at Microsoft.DotNet.UpgradeAssistant.MSBuild.PackageLoader.DoesPackageSupportTargetFrameworksAsync(NuGetReference packageReference, IEnumerable`1 targetFrameworks, CancellationToken token) in /_/src/components/Microsoft.DotNet.UpgradeAssistant.MSBuild/PackageLoader.cs:line 59
   at Microsoft.DotNet.UpgradeAssistant.Steps.Packages.Analyzers.TargetCompatibilityReferenceAnalyzer.AnalyzeAsync(IProject project, PackageAnalysisState state, CancellationToken token) in /_/src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/Analyzers/TargetCompatibilityReferenceAnalyzer.cs:line 48
   at Microsoft.DotNet.UpgradeAssistant.Steps.Packages.PackageUpdaterStep.RunPackageAnalyzersAsync(IUpgradeContext context, CancellationToken token) in /_/src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/PackageUpdaterStep.cs:line 206
   at Microsoft.DotNet.UpgradeAssistant.Steps.Packages.PackageUpdaterStep.InitializeImplAsync(IUpgradeContext context, CancellationToken token) in /_/src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/PackageUpdaterStep.cs:line 89
```

**Comment by MisinformedDNA** [comment]
@mjrousos Could we either reopen this issue or create a new one?

**Comment by mjrousos** [comment]
Yep, sounds like there are some scenarios that this still isn't working for. I'll reopen while we investigate.

**Comment by mjrousos** [comment]
To give an update here, I dug a little deeper into how auth and credential providers work for NuGet. There are some useful docs [here](https://github.com/NuGet/Home/wiki/NuGet-cross-plat-authentication-plugin) for anyone interested.

The long and the short of it is that it looks like VS and older versions of NuGet/MSBuild came with a default credential provider that would be used automatically. For newer NuGet clients (dotnet.exe, apps like Upgrade Assistant that use a modern NuGet SDK), it's necessary to install a .NET Core-compatible v2 NuGet credential provider like the one @grayjeremy linked above for Azure DevOps.

Prior to my (original) fix for this issue, we weren't looking for those sorts of providers but now we are. So, I think that things should work if (a) you have an up-to-date version of Upgrade Assistant and (b) you've installed a NuGet cred provider like [this one](https://github.com/microsoft/artifacts-credprovider).

I did find that we weren't correctly enabling interactive mode for `dotnet restore` when we called it, so I've opened a follow-up PR (#512) to address that and fix some docs/diagnostics to make it clearer how this is supposed to work.

I'm going to wait to merge that PR and close this, though, until I have a chance to work with some partners and confirm that NuGet auth is working as I expect. If anyone who has run into issues is able to test out a recent Upgrade Assistant build (after installing the cred provider linked previously) and let me know how it goes in your scenarios, that would be great.
